README.md
# OTMql4Py
## Open Trading Metaquotes4 Python Bridge

### OTMql4Py - MQL4 bindings for Python
https://github.com/OpenTrading/OTMql4Py/

Based in work by Bernd Kreuss
https://sites.google.com/site/prof7bit/metatrader-python-integration

with contributions by C. Polymeris 
http://chiselapp.com/user/polymeris/repository/metatraderpy/index

**This is a work in progress - a developers' pre-release version.**
The code is quite stable now, and there should be no major file or API changes.
It works well on builds > 6xx, but the documentation of the additions to the
original code still need writing, as well as more tests and testing on
different versions. Only Python 2.7.x is supported.

The project wiki should be open for editing by anyone logged into GitHub:
**Please report any system it works or doesn't work on in the wiki:
include the Metatrader build number, the origin of the metatrader exe,
the Windows version, and the Python version and origin of the Python.**
This code in known to run under Linux Wine (1.7.x), so this project
bridges Metatrader to Python under Linux.

### Installation

For the moment there is no installer: just "git clone" or download the
zip from github.com and unzip into an empty directory. Then recursively copy
the folder MQL4 over the MQL4 folder of your Metatrader installation. It will
not overwrite any Mt4 system files.

You already should have the python27.dll and any other Python dlls
that you will call (e.g. `pythoncom27.dll` `pythoncomloader27.dll` `pywintypes.dll`)
installed into your windows system folder (e.g. `c:\windows\system32`)

You may have to set the environment variable PYTHONHOME to the root
of your Python installation (e.g. `c:\Python27` ).

### Source

The source code to generate the py27.dll are in the src directory.
The sources have minor fixes to the original code other than the py26 -> py27
syntactic conversion, and still need recompiling: the dll checked-in at the
moment is from:
http://chiselapp.com/user/polymeris/repository/metatraderpy/index


### Project

Please file any bugs in the issues tracker:
https://github.com/OpenTrading/OTMql4Py/issues

Use the Wiki to start topics for discussion:
https://github.com/OpenTrading/OTMql4Py/wiki
It's better to use the wiki for knowledge capture, and then we can pull
the important pages back into the documentation in the share/doc directory.
You will need to be signed into github.com to see or edit in the wiki.

If you know of any threads in the forums that discuss this code,
please post a message to say this project is now on github.com at
https://github.com/OpenTrading/OTMql4Py/
## OTMql4Py Notes

### Changes

OTMql4Py has some changes to how the compiled dll code is used,
and includes a Python file `OTMql427.py` with extra functionality.

#### Unicode Functions

Build 600 of Metatrader4 changed the definition of the fundamental datatype
`string` from being ASCII to being Unicode, and broke every compiled
library called by Mt4 that send or received a string. The basic change
with implemented as a minor patch release rolled out on live systems.

Metaquotes should have left the existing string definition to be ASCII,
and introduced a new `unicode` datatype, along with Unicode aware functions,
usually with the same function name, but with a `W` appended, like Windows
does. You can still use ASCII strings in Metatrader4 >600: they are simply
arrays of `uchar`, and there are conversion functions such as
`StringToCharArray`. 

 
#### Initialization Code

The initialization code `iPyInit` in `MQL4/Include/OTMql4/OTPy27.mqh`
initializes the Python environment. This should be called from your
`OnInit()` function. It is safe to call it a second time; subsequent
calls will just be ignored.

It has an integer return value, and should return 0. 
A return value of -1 is a panic: remove the expert if it requires Python.

It calls the compiled `PyInitialize()` and then imports some standard
system modules. Then it prepends the `sys.path` with the directory
`MQL4/Python`, which should have been created when you installed `OTMql4Py`.
In that directory should be a (possibly empty) file `__init__.py`, so
that you can import modules found in that directory into Python.

The intialization code will import the module `OTMql427` found in that directory
to give some added functionality. If it can't import the module `OTMql427`
it will signal a panic by returning -1: you should fix the problem before
going any futher.

#### Global Variables

The `iPyInit` initialization creates a temporary global veriable called
`fPythonUsers` and increments it by one each time it is called.
`vPyDeInit` decrements it by one each time it is called, and if `fPythonUsers`
is zero, then it calls `OTMql427.vPyDeInit` to unload the Python interpreter.
`fPythonUsers` should always be equal to the number of charts and scripts
Python is being used on. Unfortunately, if your recompile your expert
while Python is loaded, then Mt4 will deinit your expert and re-init the
expert. If you only had one chart using Python (fPythonUsers=1), then
this will unload Python when Mt4 deinits the expert, and when Mt4 re-inits
the Python, it will fail to initialize the `py27.dll` properly. If Mt4
`OnInit` had a required `reason` argument the way `OnDeinit` does, we
could work around this, but it doesn't. Suggestions welcome...

The `iPyInit` initialization also creates a persistent global veriable called
"fDebugLevel" which is used by the logging code, and it ranges from 0 to 5:
0 : quiet, 1 : +errors, 2 : +warnings, 3 = +info, 4 : +debug, 5 : +trace.

#### Added Python Functionality

In many cases, you should use `uPySafeEval` to evaluate a python
expression that will evaluate to a string and return its value. It calls
`OTMql427.sPySafeEval` in Python which wraps the code to be evaluated
in a `try:/except:` clause and catches the error. If there's an error,
the error is returned as a string, prepended with `ERROR: `.

In the caller you should have something like:
    if (StringFind(uSource, "ERROR:", 0) == 0) {
      Print("Error in Python evaluating: " + uSource + "\n" + res);
      <do something as a result of the failure>
    }


### Testing

There are some initial tests in the file
`MQL4/Scripts/OTMql4/OTMql4PyTest.mq4`
Attach this script to a chart and it will run a series of simple tests;
you can choose which tests to run as inputs when you attach the script.
Look at the Experts log window for messages; errors will start with the
word `ERROR:` and should pop up a MessageBox.


### Known Issues

During testing, we have noticed an error when you are repeatedly
initilizing and uninitializing the Python interpreter. Any call of
Python generates a `exceptions.SystemError`. You must restart
Metatrader if this happens.



### Mt4 and Python Architecture

Mt4 runs an expert attached to a chart in a thread.
Python in a Mt4 expert starts a main thread that is unique to the
application, and starts another worker thread. The main thread is blocked
until Python is unloaded, and everything takes place in the worker thread.
The worker thread can start more threads.

You can send commands from Mt4 to Python, but Python can't send commands
to Mt4. So what you can do is queue up things for Mt4 to do, and
periodically get Mt4 to ask Python if it has anything it wants Mt4 to do.
It's easiest to think of registering an EventTimer with your expert, and
to have `OnTimer` function periodically ask Python if it has any work.
(Remember that `EventTimer` events fire even if the market is closed or
you have no connection, unlike `OnTick` events.)

The problem is that Mt4 does not have an `Eval` command, so if Python wants
Mt4 to do something, like `OrderSend`, we can't just pass a string to Mt4
that says `OrderSend(...)` and expect Mt4 to eval it: there is no eval.
So what we have done is write a simplistic replacement to what should
be an `Eval` command in Mt4, that will parse as string and execute it.
See that function `zMt4LibProcessCmd` in
`MQL4/Libraries/OTMql4/OTLibMt4ProcessCmd.mq4` of the `OTMql4Lib`
package (https://github.com/OpenTrading/OTMql4Lib/).  See also
`OTLibProcessCmd.mq4` for an example of how to extend
`zMt4LibProcessCmd` to process your own functions.

This is not a proper replacement to an `Eval` function. It only executes
commands that it knows about, and has no ability to decipher complex commands.
Still, it allows Python to make decisions and ask Mt4 to do things,
which is enough for order and account processing. You can use this for
bi-directional interaction with a program outside of Mt4. For example,
you could use a call from `OnTick` into Python to send tick and bar info
to another program, and then use a call from `OnTimer` into Python to
ask if there are any requests back from the program that it wants Mt4
to evaluate.

You may also want to use threads and queues within Python to make sure
that any command that you ask Python to execute is accomplished within
the time it takes to get the next tick: programming with Python iterators
is probably a good idea. It is untested as to whether
the Python code dispatched from `OnTick` can block the chart or not.

In our demo experts, we have set the timer interval to be quite high: 10 seconds.
What you attach the expert to the chart, you can set it to what you want.
It's set high for testing; obviously if you have a lot of charts with OTMqlPy
and set the interval low, it may add a lot of work to Mt4. 
### Development

#### Python Coding Practices

##### Type Prefixing

We prefix all variable, function and method names with a lowercase letter
that indicates the type; it helps you anticipate what type a quantity
is, and to make explicit type conversions. This is not a rigid requirement,
but it helps cut down on a very common type of mistake, pun intended.

If the type changes, assign it to a new name with the initial letter changed.
In Python, this scheme has the added advantage of avoiding any naming
conflict with built-ins, or most Python standard library modules.

Choose the initial letter from the following list:

||'''Prefix Letter'''	||'''Variable Type'''					||
|| a			|| numpy array 						||
|| b			|| boolean						||
|| c			|| complex						||
|| d			|| alist or dictionary					||
|| e			|| error string return value, empty for success		||
|| f			|| double	       	      	    			||
|| g			|| generic - unknown or mutable				||
|| h			|| hypertext - HTML/XML entity encoded string		||
|| i			|| integer		       	       			||
|| l			|| list set or tuple					||
|| o			|| instance or pointer to a class			||
|| r			|| pandas series       	    				||
|| m			|| pandas dataframe    	    				||
|| p			|| pandas panel	       	    				||
|| s			|| ASCII string	      	     				||
|| t			|| date/time value					||
|| u			|| Unicode string					||
|| v			|| void - return value not to be used			||
|| z			|| non-empty string return value - empty is failure	||



CHANGELOG.txt
commit 9848b03cc14d8771e125292db3bcaffae0356170
Date:   Thu Feb 25 15:05:03 2016 +0200

    Added ePySafeExec

commit cf85929708279f01e248c3cb827c314193aa4f8f
Date:   Thu Feb 25 00:13:58 2016 +0200

    Tagged 1.1.0.1

commit e86c8f0a4b0fe0d5bd62b4b5e494754d35b358a7
Date:   Thu Feb 25 00:11:26 2016 +0200

    Fixed documentation in Creole format

commit 14cebc1a80b08940a35259eac54133879a811736
Date:   Tue Feb 23 05:15:56 2016 +0200

    Updated documentation

commit 5b0e349261a0061191514f82933736b1e9f39ea2
Date:   Mon Feb 22 21:43:52 2016 +0200

    Updated documentation

commit e35a3b704fe7567467e0462342a6feffeeb8b8e0
Date:   Mon Feb 22 21:43:36 2016 +0200

    Updated documentation

commit 8703683f472fbae91b217797e1c7be13e36ea2b8
Date:   Mon Feb 22 16:25:52 2016 +0200

    Updated documentation

commit bf957932470fd4f28c7c80c8e5b32e110145486d
Date:   Mon Feb 22 03:07:15 2016 +0200

    Updated documentation

commit 74dd3cbf144c38e3d369583876eca293d3259fa6
Date:   Sun Feb 21 06:13:03 2016 +0200

    Added more documentation in Creole format

commit fe4e64e3a8900c7bbf9bd90a5a6eba7b4eb32baf
Date:   Sat Feb 20 08:08:10 2016 +0200

    Updated documentation

commit cf7d608d6e1b1164231f509f61ddca40f0304048
Date:   Fri Feb 19 23:29:23 2016 +0200

    Fixed issue #6

commit b9492355059eb8329c59f664bb40a36c9fd388e0
Date:   Fri Feb 19 23:03:18 2016 +0200

    Added more documentation in Creole format

commit 029a8e2190d4bde31eb287abf2b9e738a8e93ac4
Date:   Fri Feb 19 22:36:14 2016 +0200

    Moved documentation to a creole format

commit 883fc8545737c8def0b453b5ad2d6082658fbf5f
Date:   Fri Feb 19 22:28:20 2016 +0200

    Moved documentation to a creole format

commit 4add71613d9b4bacfc3ba40d0d3fc22e3723a3e9
Date:   Fri Feb 19 22:12:33 2016 +0200

    Moved documentation to a wiki format

commit d13a7d21e94805a75ea431959140c1605f5c429e
Date:   Fri Aug 21 19:10:34 2015 +0300

    Wired up the global variable fDebugLevel

commit 91fef462ef339328ff90380ffaa91a9ea87a0313
Date:   Fri Aug 14 23:19:34 2015 +0300

    Fixes for OTPyTestZmqEA with OpenTrader

commit 473308903af5b8d2941cab14609caa86072c4da6
Date:   Wed Aug 12 18:20:29 2015 +0300

    Refactored to put Zmq in OpenTrader - starting testing

commit 0add369ef8a36b7ca7e0dfee0dbefa80318c479e
Date:   Mon Jul 13 20:21:41 2015 +0300

     Tried to fix communications for errors but it still needs a rewrite

commit ff8291f5e3e0c65e5e8362360bb9d51ae87bf2f9
Date:   Sun Jul 12 23:49:31 2015 +0300

    Fixed test Scripts

commit 149c308afe95397e5d65f3e17cd7e761be08eddb
Date:   Sun Jul 12 22:30:30 2015 +0300

    Minor refactoring of Mql4Chart

commit 5f97815adde4ae231ae0b6007e0a542b7fae48b2
Date:   Sun Jun 21 16:40:48 2015 +0300

    Doc updates.

commit d6f34f97eba81a38780b5a457cb2b162f5a7e7bd
Date:   Tue Jun 16 23:22:49 2015 +0300

    Minor bugfix.

commit d97d429bca359953442143124314a3dc4043d3dc
Date:   Thu Jun 4 16:42:23 2015 +0300

    Moved MQL4/Python/OTMql427/cmd2plus.py to OpenTrader
    added MQL4/Python/OTMql427/SimpleFormat.py

commit d32e66ebae7a0223709227fca056da56cc958181
Date:   Wed Jun 3 09:52:21 2015 +0300

    Minor extensions to Mq4Chart.py.

commit 3953bf485970d79165f6db5cedf5d36fc634a831
Date:   Wed May 27 12:11:55 2015 +0300

    Fleshed out share/doc/02Architecture.txt.

commit faa9763780de8434eb6eb1f7e857fb59d1b8ffff
Date:   Thu May 21 23:10:09 2015 +0300

    Added MQL4/Python/OTMql427/cmd2plus.py which is a hacked version
    of cmd2 - to support PikaCmd2.py in OTMql4AMQP.

commit b325b974eedef94eeb26618e0e2ca0461319cbaf
Date:   Mon May 18 15:39:08 2015 +0300

    Refactored out MQL4/Libraries/OTMql4/OTPyChart.mq4 for code
    common to Pika and Zmq.

commit 288780ecba5fd0a1a80b0361733c2e74365594d9
Date:   Fri May 15 11:28:53 2015 +0300

    Refactoring to pull common code from OTMql4AMQP and OTMql4Zmq into OTMql4Py.

commit 28e134e2c0bb8f4707bf299c18fe7d5b5de55795
Date:   Fri May 8 19:10:13 2015 +0300

    Minor refactoring.

commit 5241bee54c9bfe7ec626aa2d383f3667377e8d2c
Date:   Fri May 8 16:46:51 2015 +0300

    Refactored the Chart code from OTMql4Zmq and OTMql4AMQP into Mq4Chart.py.

commit d80d90d6a98a398c542287f94f880e53172d0b0f
Date:   Thu May 7 01:05:44 2015 +0300

    Broke out functions from MQL4/Include/OTMql4/OTPy27.mqh
    into a proper library file:
    MQL4/Libraries/OTMql4/OTLibPy27.mq4
    Include this in your work, and leave the other one for internal use.

commit 2bd2e2c23546b33a57aa7213a2424242ee69a243
Date:   Thu May 7 01:03:33 2015 +0300

    Broke out some of the functionity of
    MQL4/Python/OTMql427/__init__.py
    into
    MQL4/Python/OTMql427/Mt4SafeEval.py
    MQL4/Python/OTMql427/Mt4Tkinter.py

commit f5e046f003ba2f4ef63ab7297933d341278e47b9
Date:   Fri May 1 17:50:16 2015 +0300

    Moved MQL4/Experts/OTMql4/OTPyTestZmqEA.mq4 to OTMql4Zmq

commit 43b2f2289e28836bd82b3918a4ce106ec548a799
Date:   Tue Apr 28 16:36:06 2015 +0300

    Moved share/doc/05PyZmq.txt to
    https://github.com/OpenTrading/OTMql4Zmq

commit 4f673ad5d60d3e2c7caaa85c797cfca45aa328cd
Date:   Tue Apr 28 12:38:42 2015 +0300

    Moved MQL4/Python/OTMql427/ZmqChart.py to
    https://github.com/OpenTrading/OTMql4Zmq/
    so now OTMql4Zmq will be both the C-coded and Python Zmq.

commit 059e5a33bc7e7386fef2aba07453941361c89c34
Date:   Sun Apr 26 22:38:43 2015 +0300

    Cleaned things up for pylint. The rcfile that we use
    is checked in at etc/pylint.rc.

commit 15053ebb820a9b7953acbb0a1aee8649d4e85b20
Date:   Fri Apr 24 18:13:13 2015 +0300

    Introduced the Python lgging functions - still need testing.
    
    Wrote up some documentation on bi-directional architecture for
    Python and Mt4.

commit 26337dfe123a7adb7fb60104cd3d560c45c3b808
Date:   Wed Apr 22 14:44:21 2015 +0300


MQL4/Python/OTMql427.py
# -*-mode: python; fill-column: 75; tab-width: 8; coding: utf-8-dos -*-

import sys, os, thread, threading

# import rpdb2; rpdb2.start_embedded_debugger('foobar')
    
import atexit
import logging
oLOG=logging
atexit._exithandlers=atexit._exithandlers[:-1]
def vShutupShutdown(*args, **dArgs): pass
oLOG.shutdown=vShutupShutdown

def sPySafeEval(sPyCode):
    dGlobals=sys.modules['__main__'].__dict__
    s = "try:\n    sRetval=" + sPyCode + "\nexcept Exception,e:\n    sRetval='ERROR: '+str(e)"
    try:
        k = compile(s, '<string>', 'exec')
    except Exception, e:
        sRetval="ERROR: Python error compiling " + sPyCode+ ': '+str(e)
        sys.exc_clear()
        return sRetval
    
    try:
        eval(k, dGlobals, dGlobals)
        if dGlobals['sRetval']:
            sRetval=str(dGlobals['sRetval'])
        else:
            sRetval=""
    except Exception, e:
        sRetval="ERROR: Python error evaling " + sPyCode+ ': '+str(e)
        sys.exc_clear()
        return sRetval
    
    if sRetval.find('ERROR:') == 0:
        sys.exc_clear()
    return sRetval



try:
    import win32ui, win32con

    def iMessageBox (sMsg, sTitle, iType, iIcon):
        i=win32ui.MessageBox(sMsg, sTitle, iType | iIcon)
        # while != 0 ?
        win32ui.PumpWaitingMessages()
        return i
except ImportError:
    def iMessageBox (sMsg, sTitle, iType, iIcon):
        return -1

# from win32con
MB_OK = 0
MB_OKCANCEL = 1
MB_ABORTRETRYIGNORE = 2
MB_YESNOCANCEL = 3
MB_YESNO = 4
MB_RETRYCANCEL = 5
MB_ICONHAND = 16
MB_ICONQUESTION = 32
MB_ICONEXCLAMATION = 48
MB_ICONASTERISK = 64
MB_ICONWARNING = MB_ICONEXCLAMATION
MB_ICONERROR = MB_ICONHAND
MB_ICONINFORMATION = MB_ICONASTERISK
MB_ICONSTOP = MB_ICONHAND
IDOK = 1
IDCANCEL = 2
IDABORT = 3
IDRETRY = 4
IDIGNORE = 5
IDYES = 6
IDNO = 7
IDCLOSE = 8
IDHELP = 9

def vLog(iLevel, sMsg, *lArgs, **dKeys):
    """
    Level                                Numeric value
    ------                               -----
    CRITICAL                             50
    ERROR                                40
    WARNING                              30
    INFO                                 20
    DEBUG                                10
    NOTSET                               0

From OFLibLogging.mqh:
#define LOG_PANIC 0 // unused
#define LOG_ERROR 1
#define LOG_WARN 2
#define LOG_INFO 3
#define LOG_DEBUG 4
#define LOG_TRACE 5
#define LOG_MAX 5
    """
    import threading, thread
    iId=thread.get_ident()
    lLogArray=["CRITICAL", "ERROR", "WARNING", "INFO", "DEBUG", "TRACE"]
    try:
        assert 0 <= int(iLevel) <= 5
    except:
        iLevel=4
    try:
        # oLOG.log(50 - iLevel*4, sMsg)
        if iLevel <= 0:
            # oLOG.critical(sMsg, *lArgs, **dKeys)
            print str(iId) + " " +lLogArray[iLevel], sMsg
        elif iLevel <= 1:
            # oLOG.error(sMsg, *lArgs, **dKeys)
            print str(iId) + " " +lLogArray[iLevel], sMsg
        elif iLevel <= 2:
            # oLOG.warning(sMsg, *lArgs, **dKeys)
            print str(iId) + " " +lLogArray[iLevel], sMsg
        elif iLevel <= 3:
            # oLOG.info(sMsg, *lArgs, **dKeys)
            print str(iId) + " " +lLogArray[iLevel], sMsg
        else:
            # oLOG.debug(sMsg, *lArgs, **dKeys)
            print str(iId) + " " +lLogArray[4], sMsg
        sys.stdout.flush()
        # ValueError when logging to a closed file
    except IOError:
        # [Errno 28] No space left on device - is that IOError?
        # probably filled the disk
        pass
    except Exception, e:
        print str(iId) + " ERROR: in vLog ", str(e), iLevel, sMsg
        sys.exc_clear()
        pass

__root__=None
__outfile__=None

def bStartTkinter():
    global __root__, __outfile__
    import Tkinter
    if __root__ is None:
        if not hasattr(sys, 'argv'):
            # __file__
            sys.argv=['py27.py']
        try:
            # should start this in a thread and leave it running?
            __root__=Tkinter.Tk()
            __root__.withdraw()
        except Exception ,e:
            sShowError('bStartTkinter', "Error starting Tkinter\n%s" % (e,))
            return False

        if 1:
            sShowInfo('bStartTkinter', "Started Tkinter")
        else:
            # always pegs to CPU at max
            try:
                import TkFileIO
                # _default_root
                oTop=__root__
                frame = TkFileIO.MainFrame(oTop, title ='Python IO')
                oTop.deiconify()
                oTop.tkraise()
                # i=oTop.tk.dooneevent()
                oTop.update()
                sys.stdout=frame.shell
                sys.stderr=frame.shell
                print "INFO: bStartTkinter: Started TkFileIO"
                oTop.update()
            except Exception ,e:
                sShowError('bStartTkinter', "Error starting TkFileIO\n%s" % (e,))
                # return False
        
        return True

    elif __root__.__class__ == Tkinter.Tk:
        sShowInfo('bStartTkinter', "Already Started Tkinter")
        return True
    else:
        sShowError('bStartTkinter', "Error starting Tkinter - %s" % (
            __root__.__class__,))
        return False        
        
def sShowError(sTitle, sMsg):
    global __root__
    if __root__:
        import tkMessageBox
        tkMessageBox.showerror(sTitle, sMsg)
    elif __outfile__:
        __outfile__.write('ERROR : ' + sTitle + ' ' + sMsg + '\n')
        
def sShowInfo(sTitle, sMsg):
    global __root__
    if __root__ is not None:
        import tkMessageBox
        tkMessageBox.showinfo(sTitle, sMsg)
    elif __outfile__ is not None:
        __outfile__.write('INFO : ' + sTitle + ' ' + sMsg + '\n')

def bStartFile(sStdout):
    global __outfile__

    # __outfile__ is None:
    if sStdout == "":
        pass
    elif not __outfile__:
        try:
            sStdout=os.path.join(os.path.dirname(__file__), sStdout)
            __outfile__ = open(sStdout, 'w', 1)
            sys.stdout = sys.stderr = __outfile__
            assert sys.stdout != sys.__stdout__

            print sys.version
            sys.stdout.flush()
            assert os.path.isfile(sStdout)
        except Exception, e:
            # may be in trouble logging here if stdout was not opened
            sys.exc_clear()
            return False
        
        try:
            # oLOG.basicConfig(level=logging.DEBUG)
            print "vPyInit - Opened %s %s" % (__outfile__.name, sStdout,)
            print "vPyInit - Thread " + threading.currentThread().getName() + \
                  " number " + str(thread.get_ident())
            return True
        except Exception,e:
            print "vPyInit - Error opening %s\n%s" % (sStdout, str(e),)
            sys.exc_clear()
            return False

    elif __outfile__:
        # __outfile__.seek(0, 0)
        # __outfile__.truncate(0)
        return True
    
def ePyInit(sStdout):
    vPyInit(sStdout)
    return ""

def vPyInit(sStdout):
    bStartFile(sStdout)
    # bStartTkinter()

def vPyDeInit():
    global __root__, __outfile__
    if __outfile__:
        try:
            sys.stdout = sys.__stdout__
            sys.stderr = sys.__stderr__
            sName=__outfile__.name
            # sShowInfo('vPyDeInit', "Closing %s" % (sName,))
            __outfile__.write('INFO : vPyDeInit ' + "Closing outfile %s\n" % (sName,))
            # oLOG.shutdown()
            __outfile__.flush()
            __outfile__.close()
            __outfile__=None
        except Exception,e:
            # You probably have not stdout so no point in logging it!
            print "Error closing %s\n%s" % (__outfile__, str(e),)
            sys.exc_clear()
        
    if __root__:
        __root__.destroy()
        __root__=None

    sys.exc_clear()

def test():
    # changing sys.stdout can't be done under doctest
    if os.path.isfile('test.txt'): os.remove('test.txt')

    vPyInit('test.txt')
    assert os.path.isfile('test.txt')

    vPyInit('test.txt')
    assert os.path.isfile('test.txt')
    assert sys.stdout != sys.__stdout__
    assert sys.stderr != sys.__stderr__

    # This can't be done under doctest
    s=sPySafeEval('foobar')
    assert s.find('ERROR:') == 0
    assert s == "ERROR: name 'foobar' is not defined"

    vLog(0, "Level 0")
    vLog(1, "Level 1")
    vLog(2, "Level 2")
    vLog(3, "Level 3")
    vLog(4, "Level 4")

    # oLOG.root.setLevel(40)
    # oLOG.debug("NONONO oLOG.debug")
    # vLog(4, "NONO vLog 4")

    vPyDeInit()
    assert sys.stdout == sys.__stdout__
    assert sys.stderr == sys.__stderr__
    assert __root__ is None
    assert __outfile__ is None

    # should check contents of test.txt
    oFd=open('test.txt', 'r')
    sContents=oFd.read()
    oFd.close()
    assert sContents.find("Level 4") > 0
    assert sContents.find("NONO") < 0

def vTestTkinter():
    if 0:
        bStartTkinter()
        assert __root__
        sShowError('Foo','Bar')

def vTestMessageBox():
    i = iMessageBox("Hi there", "Yes No Cancel",
                    MB_YESNOCANCEL, MB_ICONINFORMATION)
    if i == IDCANCEL:
        pass
    elif i == IDYES:
        pass
    elif i == IDOK:
        pass
    # win32ui.PumpWaitingMessages()

